test/test_sparse.py

```diff
@@ -143,6 +143,62 @@ def f5():
             # Check warn-once:
             self.assertEqual(len(cm.warnings), 1)
 
+    @parametrize('fast_mode', (True, False))
+    def test_gradcheck_check_sparse_nnz(self, fast_mode):
+        """Tests for deprecated check_sparse_nnz keyword argument of gradcheck.
+
+        Suggested deprecation steps:
+        2.1: Specification of check_sparse_nnz triggers a warning.
+        2.2: Specification of check_sparse_nnz triggers a warning and
+             an exception. Update gradcheck and this test accordingly.
+        2.3: Specification of check_sparse_nnz triggers an
+             exception. Remove all check_sparse_nnz usages from
+             gradcheck and delete this test.
+
+        """
+        def fn(x, masked_grad):
+            return x.to_dense(masked_grad=masked_grad)
+
+        def test(x, masked_grad, masked, check_sparse_nnz):
+            x = x.detach().clone().requires_grad_()
+            torch.autograd.gradcheck(fn, (x, masked_grad), masked=masked, check_sparse_nnz=check_sparse_nnz, fast_mode=fast_mode)
+
+        x = torch.tensor([[0, 2], [3, 4]], dtype=torch.float64).to_sparse()
+
+        for masked_grad, masked, check_sparse_nnz in itertools.product(*[(True, False, None)] * 3):
+            effective_masked_grad = True if masked_grad is None else masked_grad
+            effective_check_sparse_nnz = False if check_sparse_nnz is None else check_sparse_nnz
+            # For BC, the effective masked depends on the value of specified check_sparse_nnz:
+            effective_masked = (check_sparse_nnz if check_sparse_nnz is not None else False) if masked is None else masked
+
+            warn_using_check_sparse_nnz = self.assertWarns(
+                UserWarning,
+                msg=('Backwards compatibility: check_sparse_nnz is deprecated, it will be removed in a future version of PyTorch.'
+                     f' Use masked={effective_check_sparse_nnz} instead.'))
+            raise_on_non_equal_masked_and_check_sparse_nnz = self.assertRaisesRegex(
+                ValueError,
+                f"Expected specified check_sparse_nnz [(]={effective_check_sparse_nnz}[)]"
+                f" to be equal to masked [(]={effective_masked}[)]")
+            raise_jacobian_mismatch = self.assertRaisesRegex(RuntimeError, "Jacobian mismatch for output 0 with respect to input 0")
+
+            def run_test():
+                if effective_masked_grad != effective_masked and not fast_mode:
```

Comment:
noob question: I would expect the test to pass if effective_masked_grad=False and effective_masked=True, since passing masked=True makes gradcheck strictly more lenient right (you're checking fewer elements). What am I missing?

Reply:
Consider the following example:
```
>>> x = torch.tensor([[0, 2], [3, 4]], dtype=torch.float64).to_sparse().requires_grad_()
>>> x
tensor(indices=tensor([[0, 1, 1],
                       [1, 0, 1]]),
       values=tensor([2., 3., 4.]),
       size=(2, 2), nnz=3, dtype=torch.float64, layout=torch.sparse_coo,
       requires_grad=True)
>>> torch.autograd.gradcheck(lambda x: x.to_dense(masked_grad=False), x, masked=True)
<snip>
    raise GradcheckError(_get_notallclose_msg(a, n, i, j, complex_indices, test_imag))
torch.autograd.gradcheck.GradcheckError: Jacobian mismatch for output 0 with respect to input 0,
numerical:tensor([[0.0000, 0.0000, 0.0000, 0.0000],
        [0.0000, 1.0000, 0.0000, 0.0000],
        [0.0000, 0.0000, 1.0000, 0.0000],
        [0.0000, 0.0000, 0.0000, 1.0000]], dtype=torch.float64)
analytical:tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]], dtype=torch.float64)

>>> torch.autograd.gradcheck(lambda x: x.to_dense(masked_grad=True), x, masked=True)
True
```
That is, the `masked` kw argument to gradcheck affects the calculation of a numerical jacobian while the `masked` kw argument to `to_dense` method affects the calculation of an analytical jacobian. To match numerical and analytical jacobians, the `masked` and `masked_grad` keyword arguments must be consistent, otherwise, Jacobian mismatch exception will be expected.

Reply:
Makes sense, thanks!